src/logfile_values.py

```diff
@@ -7,42 +7,271 @@
+
+And aggregate data by time period with different functions:
+
+median, mean, sum, min, max, count, frequency, speed (??)
+
+count_100 - counts values > 100
+count_100_percentage - estimates percentage of values > 100
+
+python logfile_values.py  --metric "metric1:1:mean:1d" --metric "metric2:3:count:60s"
+
 """
+import re
+import time
+import os
+import gzip
+import logging
 
 from argparse import ArgumentParser, ArgumentTypeError
-from time import time
+from datetime import timedelta
 
 
 class Metric:
     def __init__(self, arg):
         if ':' not in arg:
             raise ArgumentTypeError('Argument must have ":"')
-        self.name, column = arg.split(':', 1)
-        if not column.isdecimal():
-            raise ArgumentTypeError('Column must be a number')
-        self.column = int(column)
+        parts = arg.split(':')
+        if len(parts) != 4 and len(parts) != 2:
+            raise ValueError('Wrong number of options')
+
+        # Accepts 2 or 4 options. If 2 then function and period are NULL
+        parts += [None] * (4 - len(parts))
+        self.name, column, function, period = parts
+        if period:
+            pattern = re.compile('^\d+[A-Za-z]{1,3}$')
+            if not pattern.match(period):
+                raise ArgumentTypeError('Period must have number and unit')
+
+        self.column = column
+        self.function = function
+        self.period = period
+        self.values = []  # Container for metric values
+        self.last_value = 0
+        self.now = int(time.time())
+
+    def get_timeshift(self):
+        if self.period:
+            units = [
+                ('s', 1),
+                ('min', 60),
+                ('h', 60 * 60),
+                ('d', 60 * 60 * 24),
+            ]
+            for unit, mul in units:
+                if self.period.lower().endswith(unit):
+                    return int(self.period[:-len(unit)].strip()) * mul
+        return 0
+
+    def estimate_columns_value(self, fields):
+        '''
+        Apply some arithmetic on several columns if needed
+        Warning: Estimates regardless of arithmetics rules        
+        '''
+        arr = [
+            s for s in self.column if s.isdigit() or s in ['/', '*', '+', '-']
+        ]
+        result = 0
+        for index, value in enumerate(arr):
+            if value.isdigit() and result == 0:
+                result += float(fields[int(value)])
+            elif value == '/':
+                result = result / float(fields[int(arr[index + 1])])
+            elif value == '*':
+                result = result * float(fields[int(arr[index + 1])])
+            elif value == '+':
+                result = result + float(fields[int(arr[index + 1])])
+            elif value == '-':
+                result = result - float(fields[int(arr[index + 1])])
+        return result
+
+    def get_median(self):
+        sorted_list = sorted(self.values)
+        i = len(sorted_list)
+        if not i % 2:
```

Comment:
I just noticed this is wrong.  I don't think `not` should be there.

Reply:
why 
not means the number of elements is odd (reminder == 0), so we try to get one number above and below from center and divide them by 2.
If it is not odd then just taking the middle number

Reply:
True, all good.  I am sorry for confusion.